commit 796d3ed23771a922b7185af1516c9c2a5f1527cf
Author: Karsten Sperling <ksperling@apple.com>
Date:   Thu Feb 5 10:09:50 2026 +1300

    [posix] Add uart-exclusive option to enable flock / TIOCEXCL

diff --git a/third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp b/third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp
index afe0842ef..bfedc9858 100644
--- a/third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp
+++ b/third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp
@@ -49,6 +49,7 @@
 #endif
 #include <stdarg.h>
 #include <stdlib.h>
+#include <sys/file.h>
 #include <sys/ioctl.h>
 #include <sys/resource.h>
 #include <sys/stat.h>
@@ -459,6 +460,27 @@ int HdlcInterface::OpenFile(const Url::Url &aRadioUrl)
         ExitNow();
     }
 
+    if (aRadioUrl.HasParam("uart-exclusive"))
+    {
+        // Lock the device early to prevent concurrent access
+        if (flock(fd, LOCK_EX | LOCK_NB) == -1)
+        {
+            // LogCrit("flock failed: %s (device may already be in use)", strerror(errno));
+            perror("flock uart failed, device already in use");
+            close(fd);
+            fd = -1;
+            ExitNow();
+        }
+
+#ifdef TIOCEXCL
+        // Set exclusive access mode if supported by the platform
+        if (ioctl(fd, TIOCEXCL) == -1)
+        {
+            LogWarn("ioctl(TIOCEXCL) failed: %s", strerror(errno));
+        }
+#endif
+    }
+
     if (isatty(fd))
     {
         struct termios tios;
diff --git a/third_party/openthread/repo/src/posix/platform/radio_url.cpp b/third_party/openthread/repo/src/posix/platform/radio_url.cpp
index 0a0cf73a6..7d6653e00 100644
--- a/third_party/openthread/repo/src/posix/platform/radio_url.cpp
+++ b/third_party/openthread/repo/src/posix/platform/radio_url.cpp
@@ -85,6 +85,7 @@ const char *otSysGetRadioUrlHelpString(void)
     "    uart-flow-control              Enable flow control, disabled by default.\n"                 \
     "    uart-init-deassert             Deassert lines on init when flow control is disabled.\n"     \
     "    uart-reset                     Reset connection after hard resetting RCP(USB CDC ACM).\n"   \
+    "    uart-exclusive                 Lock uart device using flock / TIOCEXCL.\n"   \
     "\n"
 #else
 #define OT_SPINEL_HDLC_RADIO_URL_HELP_BUS
